sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveInspectors.scala

```diff
@@ -479,8 +340,299 @@ private[hive] trait HiveInspectors {
   }
 
   /**
-   * Builds specific unwrappers ahead of time according to object inspector
+   * Builds unwrappers ahead of time according to object inspector
    * types to avoid pattern matching and branching costs per row.
+   *
+   * Strictly follows the following order in unwrapping (constant OI has the higher priority):
+   * Constant Null object inspector =>
+   *   return null
+   * Constant object inspector =>
+   *   extract the value from constant object inspector
+   * If object inspector prefers writable =>
+   *   extract writable from `data` and then get the catalyst type from the writable
+   * Extract the java object directly from the object inspector
+   *
+   * NOTICE: the complex data type requires recursive unwrapping.
+   *
+   * @param objectInspector the ObjectInspector used to create an unwrapper.
+   * @return A function that unwraps data objects.
+   *         Use the overloaded HiveStructField version for in-place updating of a MutableRow.
+   */
+  def unwrapperFor(objectInspector: ObjectInspector): Any => Any =
+    objectInspector match {
+      case coi: ConstantObjectInspector if coi.getWritableConstantValue == null =>
+        data: Any => null
+      case poi: WritableConstantStringObjectInspector =>
+        data: Any =>
+          UTF8String.fromString(poi.getWritableConstantValue.toString)
+      case poi: WritableConstantHiveVarcharObjectInspector =>
+        data: Any =>
+          UTF8String.fromString(poi.getWritableConstantValue.getHiveVarchar.getValue)
+      case poi: WritableConstantHiveCharObjectInspector =>
+        data: Any =>
+          UTF8String.fromString(poi.getWritableConstantValue.getHiveChar.getValue)
+      case poi: WritableConstantHiveDecimalObjectInspector =>
+        data: Any =>
+          HiveShim.toCatalystDecimal(
+            PrimitiveObjectInspectorFactory.javaHiveDecimalObjectInspector,
+            poi.getWritableConstantValue.getHiveDecimal)
+      case poi: WritableConstantTimestampObjectInspector =>
+        data: Any => {
+          val t = poi.getWritableConstantValue
+          t.getSeconds * 1000000L + t.getNanos / 1000L
+        }
+      case poi: WritableConstantIntObjectInspector =>
+        data: Any =>
+          poi.getWritableConstantValue.get()
```

Comment:
You're right, as the contract for a `ConstantObjectInspector` is that its object "represent constant values and can return them without an evaluation" [[1](https://github.com/apache/hive/blob/master/serde/src/java/org/apache/hadoop/hive/serde2/objectinspector/ConstantObjectInspector.java)]. I will make this change.